snuba/pipeline/pipeline_delegator.py

```diff
@@ -0,0 +1,141 @@
+
+    The pipelines that are built and executed on a particular query are determined
+    by the selector function which returns a sequence of query builder IDs. The
+    first item in the list corresponds to the pipeline that will be used to ultimately
+    return a result. Hence the selector function must return a list with at least
+    one value in it.
+
+    Subsequent (if any) pipelines that are also returned by the selector function
+    will be executed in separate threads. The results of these other queries are
+    provided to the callback function if one is provided.
+
+    This component is produced by the PipelineDelegator.
+    """
+
+    def __init__(
+        self,
+        request: Request,
+        runner: QueryRunner,
+        query_pipeline_builders: QueryPipelineBuilders,
+        selector_func: SelectorFunc,
+        callback_func: Optional[CallbackFunc],
+    ):
+        self.__request = request
+        self.__runner = runner
+        self.__query_pipeline_builders = query_pipeline_builders
+        self.__selector_func = selector_func
+        self.__callback_func = callback_func
+
+    def execute(self) -> QueryResult:
+        def build_and_execute_pipeline(
+            query_pipeline_builder: QueryPipelineBuilder,
+        ) -> QueryResult:
+            return query_pipeline_builder.build_pipeline(
+                self.__request, self.__runner
+            ).execute()
+
+        def execute_pipelines() -> Iterator[Tuple[BuilderId, QueryResult]]:
+            builder_ids = self.__selector_func(self.__request.query)
+            primary_builder_id = builder_ids.pop(0)
+
+            futures = [
+                (
+                    builder_id,
+                    executor.submit(
+                        build_and_execute_pipeline,
+                        query_pipeline_builder=self.__query_pipeline_builders[
+                            builder_id
+                        ],
+                    ),
+                )
+                for builder_id in builder_ids
+            ]
+
+            yield primary_builder_id, build_and_execute_pipeline(
+                self.__query_pipeline_builders[primary_builder_id]
+            )
+
+            yield from [
+                (builder_id, future.result()) for (builder_id, future) in futures
+            ]
+
+        generator = execute_pipelines()
+
+        query_results: QueryResults = []
+
+        try:
+            builder_id, query_result = next(generator)
+            query_results.append((builder_id, query_result))
+            return query_result
+        finally:
+
+            def execute_callback() -> None:
+                try:
+                    for builder_id, query_result in generator:
+                        query_results.append((builder_id, query_result))
+
+                    if self.__callback_func is not None:
+                        self.__callback_func(query_results)
+                except QueryException as error:
```

Comment:
What happens for other Exceptions? Like query validation errors (they can still happen during processing and translation). Those are not QueryExceptions.

Reply:
It would be nice to have some tests around these error cases as well.

Reply:
Updated to handle any exception